### PR TITLE
ARROW-4178: [C++] Fix TSan and UBSan errors

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
Unfortunately, UBSan flags some flatbuffers usage that I haven't been able to suppress or work around. Typically when serializing a 0-size structure, it seems flatbuffers can call `memcpy` with a null pointer (which is technically undefined behaviour [as per the C++ spec](https://en.cppreference.com/w/cpp/string/byte/memcpy)). Here is a typical stack snippet:
```
/home/antoine/miniconda3/envs/pyarrow/include/flatbuffers/flatbuffers.h:617:18: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:43:28: note: nonnull attribute specified here
    #0 0x7f6d637ba4e7 in flatbuffers::vector_downward::push(unsigned char const*, unsigned long) /home/antoine/miniconda3/envs/pyarrow/include/flatbuffers/flatbuffers.h:617:5
    #1 0x7f6d637b9cea in flatbuffers::FlatBufferBuilder::PushBytes(unsigned char const*, unsigned long) /home/antoine/miniconda3/envs/pyarrow/include/flatbuffers/flatbuffers.h:813:10
    #2 0x7f6d638e4308 in flatbuffers::Offset<flatbuffers::Vector<org::apache::arrow::flatbuf::Block const*> > flatbuffers::FlatBufferBuilder::CreateVectorOfStructs<org::apache::arrow::flatbuf::Block>(org::apache::arrow::flatbuf::Block const*, unsigned long) /home/antoine/miniconda3/envs/pyarrow/include/flatbuffers/flatbuffers.h:1208:5
    #3 0x7f6d638e235f in flatbuffers::Offset<flatbuffers::Vector<org::apache::arrow::flatbuf::Block const*> > flatbuffers::FlatBufferBuilder::CreateVectorOfStructs<org::apache::arrow::flatbuf::Block>(std::vector<org::apache::arrow::flatbuf::Block, std::allocator<org::apache::arrow::flatbuf::Block> > const&) /home/antoine/miniconda3/envs/pyarrow/include/flatbuffers/flatbuffers.h:1259:12
    #4 0x7f6d638c6e01 in arrow::ipc::internal::FileBlocksToFlatbuffer(flatbuffers::FlatBufferBuilder&, std::vector<arrow::ipc::internal::FileBlock, std::allocator<arrow::ipc::internal::FileBlock> > const&) /home/antoine/arrow/cpp/build-test/../src/arrow/ipc/metadata-internal.cc:804:14
```

**Comment by pitrou** [comment]
@xhochy @fsaintjacques This might interest you.

File: cpp/src/parquet/types.h [hunk 1]
```diff
@@ -160,7 +160,8 @@ struct ByteArray {
 };
 
 inline bool operator==(const ByteArray& left, const ByteArray& right) {
-  return left.len == right.len && 0 == std::memcmp(left.ptr, right.ptr, left.len);
+  return left.len == right.len &&
+         (left.len == 0 || std::memcmp(left.ptr, right.ptr, left.len) == 0);
```
Comment (fsaintjacques) [inline cpp/src/parquet/types.h]: In theory you'd have to limit the `memcmp` to the smallest length.

File: cpp/src/parquet/encoding-internal.h [hunk 2]
```diff
@@ -83,7 +83,10 @@ inline int DecodePlain(const uint8_t* data, int64_t data_size, int num_values,
   if (data_size < bytes_to_decode) {
     ParquetException::EofException();
   }
-  memcpy(out, data, bytes_to_decode);
+  // If bytes_to_decode == 0, data could be null
+  if (bytes_to_decode > 0) {
```
Comment (fsaintjacques) [inline cpp/src/parquet/encoding-internal.h]: Worth adding a ARROW_PREDICT_TRUE.

File: cpp/src/parquet/util/memory.cc [hunk 3]
```diff
@@ -233,8 +233,11 @@ void InMemoryOutputStream::Write(const uint8_t* data, int64_t length) {
     PARQUET_THROW_NOT_OK(buffer_->Resize(new_capacity));
     capacity_ = new_capacity;
   }
-  memcpy(Head(), data, length);
-  size_ += length;
+  // If length == 0, data may be null
+  if (length > 0) {
```
Comment (fsaintjacques) [inline cpp/src/parquet/util/memory.cc]: Worth adding a ARROW_PREDICT_TRUE.

File: cpp/src/parquet/column_reader-test.cc [hunk 4]
```diff
@@ -102,7 +102,7 @@ class TestPrimitiveReader : public ::testing::Test {
           &vresult[0] + total_values_read, &values_read));
       total_values_read += static_cast<int>(values_read);
       batch_actual += batch;
-      batch_size = std::max(batch_size * 2, 4096);
+      batch_size = std::min(1 << 24, std::max(batch_size * 2, 4096));
```
Comment (fsaintjacques) [inline cpp/src/parquet/column_reader-test.cc]: Worth adding `kMinimumBatchSize` `kMaximumBatchSize`, bonus point for `arrow::clamp`.

**Review by xhochy — APPROVED** [review]
+1, thanks looking into these tests.

File: cpp/src/arrow/compare.cc [hunk 5]
```diff
@@ -324,7 +324,15 @@ static bool IsEqualPrimitive(const PrimitiveArray& left, const PrimitiveArray& r
     right_data = right.values()->data() + right.offset() * byte_width;
   }
 
-  if (left.null_count() > 0) {
+  if (byte_width == 0) {
+    // Special case 0-width data, as the data pointers may be null
+    for (int64_t i = 0; i < left.length(); ++i) {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compare.cc]: I had forgotten about it. Too late :-/

**Pull request closed without merging**